pkg/controller/migplan/registry_health.go

```diff
@@ -0,0 +1,143 @@
+package migplan
+
+import (
+	"context"
+	"github.com/konveyor/mig-controller/pkg/apis/migration/v1alpha1"
+	"github.com/konveyor/mig-controller/pkg/compat"
+	health "github.com/konveyor/mig-controller/pkg/health"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+	event2 "sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"time"
+)
+
+
+// A registry health checker which routinely checks whether registries are healthy or not
+type registryHealth struct {
+	hostClient client.Client
+	srcClient  client.Client
+	Interval   time.Duration
+	handler    handler.EventHandler
+	queue      workqueue.RateLimitingInterface
+	predicates []predicate.Predicate
+}
+
+// Start the health checks
+func (r *registryHealth) Start(
+	handler handler.EventHandler,
+	queue workqueue.RateLimitingInterface,
+	predicates ...predicate.Predicate) error {
+
+	r.handler = handler
+	r.queue = queue
+	r.predicates = predicates
+	go r.run()
+
+	return nil
+}
+
+// Enqueue a reconcile request event for migplan
+func (r *registryHealth) enqueue(plan v1alpha1.MigPlan) {
+	event := event2.GenericEvent{
+		Meta:   &plan.ObjectMeta,
+		Object: &plan,
+	}
+	for _, predicate := range r.predicates {
+		if !predicate.Generic(event) {
+			return
+		}
+	}
+
+	r.handler.Generic(event, r.queue)
+}
+
+//Run the health checks for registry pods
+func (r *registryHealth) run() {
+	//List all the migplans using the hostClient
+	//Now using srcClient for each migplan find the registry pods, if registry container is not ready then enqueue this migplan
+	//repeat all the above steps for all the plans for both clusters
+
+	for{
+		time.Sleep(r.Interval)
+		planList, err := v1alpha1.ListPlans(r.hostClient)
+		if err != nil {
+			log.Trace(err)
+			return
+		}
+
+		for _, plan := range planList{
+			//TODO avoid race condition check
+
+
+			srcCluster, err := plan.GetSourceCluster(r.hostClient)
+			if err != nil {
+				log.Trace(err)
```

Comment:
what will happen if you hit this error and the go routine returns? Would it create a situation where we are not watching? Would you want to see if the SourceCluster is in ready state before trying to get the Client on Line 88?

Reply:
Also please add a TODO before the return, once an error occurs, we loose the watch for until the time mig-controller pod restarts

Reply:
Added a Cluster ready check before moving onto the registry pods, also added a TODO